VS 2019 build fails with syntax error

Hello,

current code fails when I tried to build IfxPy on windows (64bit, clientsdk.4.10.FC9DE, python-3.8.3 and VS2019): 

```
f:\Project\IfxPy\IfxPy>python setup.py build
Detected 64-bit Python
running build
running build_py
running build_ext
building 'IfxPy' extension
c:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -If:/Developments/Python-3.8.4\include "-Ic:/Program Files/IBM Informix Client SDK\incl\cli" "-IC:\Program Files\Python38\include" "-IC:\Program Files\Python38\include" "-Ic:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\ATLMFC\include" "-Ic:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" "-Ic:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\ATLMFC\include" "-Ic:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /Tcifxpyc.c /Fobuild\temp.win-amd64-3.8\Release\ifxpyc.obj
ifxpyc.c
ifxpyc.c(70): error C2061: syntax error: identifier 'gSmartTriggerRegister'
ifxpyc.c(70): error C2059: syntax error: ';'
ifxpyc.c(70): error C2059: syntax error: '['
ifxpyc.c(71): error C2061: syntax error: identifier 'gopenSmartTrigger'
ifxpyc.c(71): error C2059: syntax error: ';'
ifxpyc.c(72): error C2061: syntax error: identifier 'gJoinSmartTrigger'
ifxpyc.c(72): error C2059: syntax error: ';'
ifxpyc.c(212): warning C4133: '=': incompatible types - from 'storeSessionID *' to 'storeSessionID *'
ifxpyc.c(392): warning C4047: '!=': 'SQLHANDLE' differs in levels of indirection from 'int'
ifxpyc.c(593): error C2065: 'gJoinSmartTrigger': undeclared identifier
ifxpyc.c(593): error C2224: left of '.callback' must have struct/union type
ifxpyc.c(594): error C2065: 'gJoinSmartTrigger': undeclared identifier
ifxpyc.c(594): error C2224: left of '.joinSessionID' must have struct/union type
ifxpyc.c(595): error C2065: 'gJoinSmartTrigger': undeclared identifier
ifxpyc.c(595): error C2224: left of '.ControlBackToApplication' must have struct/union type
ifxpyc.c(597): error C2065: 'gJoinSmartTrigger': undeclared identifier
ifxpyc.c(720): warning C4133: '=': incompatible types - from 'storeSessionID *' to 'storeSessionID *'
ifxpyc.c(833): error C2065: 'gopenSmartTrigger': undeclared identifier
ifxpyc.c(833): error C2224: left of '.isDetachable' must have struct/union type
ifxpyc.c(834): error C2065: 'gopenSmartTrigger': undeclared identifier
ifxpyc.c(834): error C2224: left of '.maxPendingOperations' must have struct/union type
ifxpyc.c(835): error C2065: 'gopenSmartTrigger': undeclared identifier
ifxpyc.c(835): error C2224: left of '.maxRecsPerRead' must have struct/union type
ifxpyc.c(836): error C2065: 'gopenSmartTrigger': undeclared identifier
ifxpyc.c(836): error C2224: left of '.timeOut' must have struct/union type
ifxpyc.c(838): error C2065: 'gopenSmartTrigger': undeclared identifier
ifxpyc.c(890): warning C4133: '=': incompatible types - from 'storeSessionID *' to 'storeSessionID *'
ifxpyc.c(1137): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1137): error C2109: subscript requires array or pointer type
ifxpyc.c(1140): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1140): error C2109: subscript requires array or pointer type
ifxpyc.c(1143): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1143): error C2109: subscript requires array or pointer type
ifxpyc.c(1146): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1146): error C2109: subscript requires array or pointer type
ifxpyc.c(1149): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1149): error C2109: subscript requires array or pointer type
ifxpyc.c(1152): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1152): error C2109: subscript requires array or pointer type
ifxpyc.c(1155): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1155): error C2109: subscript requires array or pointer type
ifxpyc.c(1158): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1158): error C2109: subscript requires array or pointer type
ifxpyc.c(1161): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1161): error C2109: subscript requires array or pointer type
ifxpyc.c(1164): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1164): error C2109: subscript requires array or pointer type
ifxpyc.c(1172): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1172): error C2109: subscript requires array or pointer type
ifxpyc.c(1173): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1173): error C2109: subscript requires array or pointer type
ifxpyc.c(1174): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1174): error C2109: subscript requires array or pointer type
ifxpyc.c(1175): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1175): error C2109: subscript requires array or pointer type
ifxpyc.c(1176): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1176): error C2109: subscript requires array or pointer type
ifxpyc.c(1177): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1177): error C2109: subscript requires array or pointer type
ifxpyc.c(1178): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1178): error C2109: subscript requires array or pointer type
ifxpyc.c(1179): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1179): error C2109: subscript requires array or pointer type
ifxpyc.c(1181): error C2065: 'gSmartTriggerRegister': undeclared identifier
ifxpyc.c(1181): error C2109: subscript requires array or pointer type
ifxpyc.c(1181): warning C4022: 'SQLSetStmtAttrW': pointer mismatch for actual parameter 3
ifxpyc.c(1181): error C2198: 'SQLSetStmtAttrW': too few arguments for call
ifxpyc.c(3036): warning C4133: '=': incompatible types - from 'storeSessionID *' to 'storeSessionID *'
ifxpyc.c(5698): warning C4996: 'PyUnicode_GetSize': deprecated in 3.3
ifxpyc.c(5926): warning C4244: 'function': conversion from 'SQLLEN' to 'SQLSMALLINT', possible loss of data
ifxpyc.c(5926): warning C4133: 'function': incompatible types - from 'SQLLEN *' to 'SQLINTEGER *'
ifxpyc.c(6068): warning C4996: 'PyUnicode_GetSize': deprecated in 3.3
ifxpyc.c(7404): warning C4244: 'function': conversion from 'SQLLEN' to 'long', possible loss of data
ifxpyc.c(7703): warning C4244: 'function': conversion from 'SQLLEN' to 'long', possible loss of data
ifxpyc.c(7978): warning C4244: 'function': conversion from 'SQLULEN' to 'long', possible loss of data
ifxpyc.c(8282): warning C4244: 'function': conversion from 'SQLLEN' to 'long', possible loss of data
ifxpyc.c(9262): warning C4133: 'function': incompatible types - from 'SQLWCHAR *' to 'const char *'
ifxpyc.c(11245): warning C4133: 'function': incompatible types - from 'SQLUINTEGER *' to 'SQLULEN *'
ifxpyc.c(11266): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
ifxpyc.c(11286): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
ifxpyc.c(11541): warning C4133: 'function': incompatible types - from 'SQLINTEGER *' to 'SQLLEN *'
ifxpyc.c(11711): warning C4244: 'function': conversion from 'SQLLEN' to 'long', possible loss of data
error: command 'c:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.26.28801\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2
```

--- a/IfxPy/ifxpyc.c
+++ b/IfxPy/ifxpyc.c
@@ -64,12 +64,14 @@ static PyObject *persistent_list;
 // True global resources - no need for thread safety here 
 static struct _IfxPy_globals *IfxPy_globals;
 
+#ifdef HAVE_SMARTTRIGGER
 // ********** Smart Trigger Start ********************** //
 
 short static gCounter = 0;
 IFMX_REGISTER_SMART_TRIGGER  gSmartTriggerRegister[NUM_OF_SMART_TRIGGER_REGISTRATION];
 IFMX_OPEN_SMART_TRIGGER      gopenSmartTrigger;
 IFMX_JOIN_SMART_TRIGGER      gJoinSmartTrigger;
+#endif  // HAVE_SMARTTRIGGER
 
 static PyObject *my_callback[NUM_OF_SMART_TRIGGER_REGISTRATION];
 
@@ -511,6 +513,7 @@ static void _python_IfxPy_check_sql_errors(SQLHANDLE handle, SQLSMALLINT hType, 
     }
 }
 
+#ifdef HAVE_SMARTTRIGGER
 static PyObject *IfxPy_join_smart_trigger_session(PyObject *self, PyObject *args)
 {
 	PyObject *result = NULL;
@@ -1225,6 +1228,7 @@ static PyObject *IfxPy_register_smart_trigger_loop(PyObject *self, PyObject *arg
 {
 	return _python_IfxPy_register_smart_trigger_helper(self, args, TRUE);
 }
+#endif  // HAVE_SMARTTRIGGER
 
 /*    static int _python_IfxPy_assign_options( void *handle, int type, long opt_key, PyObject *data ) */
 static int _python_IfxPy_assign_options(void *handle, int type, long opt_key, PyObject *data)
@@ -12160,13 +12164,14 @@ static PyMethodDef IfxPy_Methods[] = {
     { "tables", (PyCFunction)IfxPy_tables, METH_VARARGS, "Returns a result set listing the tables and associated metadata in a database" },
     { "get_last_serial_value", (PyCFunction)IfxPy_get_last_serial_value, METH_VARARGS, "Returns last serial value inserted for identity column" },
     { "SpeedTestWithCPrimeCount", (PyCFunction)SpeedTestWithCPrimeCount, METH_VARARGS, "Used for Speed Test C & Python function by counting number of prime numbers between X and Y" },
+#ifdef HAVE_SMARTTRIGGER
     { "open_smart_trigger", (PyCFunction)IfxPy_open_smart_trigger, METH_VARARGS, "Open Smart Trigger session" },
 	{ "get_smart_trigger_session_id", (PyCFunction)IfxPy_get_smart_trigger_sessionID, METH_VARARGS, "Get already opened Smart Trigger session ID" },
 	{ "join_smart_trigger_session", (PyCFunction)IfxPy_join_smart_trigger_session, METH_VARARGS, "Join already opened Smart Trigger session ID" },
 	{ "register_smart_trigger_loop", (PyCFunction)IfxPy_register_smart_trigger_loop, METH_VARARGS, "Open Smart Trigger session with loop handled by Python" },
 	{ "register_smart_trigger_no_loop", (PyCFunction)IfxPy_register_smart_trigger_no_loop, METH_VARARGS, "Open Smart Trigger session with loop handled by application" },
 	{ "delete_smart_trigger_session", (PyCFunction)IfxPy_delete_smart_trigger_session, METH_VARARGS, "Delete opened Smart Trigger session" },
-
+#endif  // HAVE_SMARTTRIGGER
     // An end-of-listing sentinel:
     { NULL, NULL, 0, NULL }
 };

--- a/IfxPy/setup.py
+++ b/IfxPy/setup.py
@@ -42,6 +42,12 @@ machine_bits =  8 * struct.calcsize("P")
 is64Bit = True
 csdk_home = os.environ['CSDK_HOME']
 py_home = os.environ['MY_PY_DIR']
+definitions = [('HAVE_SMARTTRIGGER', None)]  # available from CSDK 4.50
+
+if "--disable_smart_triggers" in sys.argv:
+    definitions.remove(('HAVE_SMARTTRIGGER', None))
+    sys.stdout.write("Smart triggers disabled.\n")
+    sys.argv.remove("--disable_smart_triggers")
 
 if machine_bits == 64:
     is64Bit = True
@@ -54,12 +60,14 @@ if('win32' in sys.platform):
     IfxPyNative_ext_modules = Extension('IfxPy',
         include_dirs = [py_home + '\\include', csdk_home + '\\incl\\cli'],
         libraries = ['iclit09b'],
+        define_macros=definitions,
         library_dirs = [ py_home + '\libs', csdk_home + '\lib'],
         sources = ['ifxpyc.c'])
 else:
     IfxPyNative_ext_modules = Extension('IfxPy',
         include_dirs = [ py_home,  py_home + '/Include', csdk_home +'/incl/cli'],
         libraries = ['ifdmr', 'thcli'],
+        define_macros=definitions,
         library_dirs = [ csdk_home + '/lib/cli', py_home + '/Lib'],
         sources = ['ifxpyc.c'])
 

--- a/LocalBuildWindows.md
+++ b/LocalBuildWindows.md
@@ -90,6 +90,8 @@ python setup.py build > out.txt 2>&1
 # C:\work\IfxPy\IfxPy\build\lib.win-amd64-3.6\IfxPy.cp36-win_amd64.pyd
 ```
 
+If your CSDK version is 4.10 or older, use build option `--disable_smart_triggers` to avoid build failure.
+
 
 #### Other Build Options
 -------------------------

